executor/executor.go

```diff
@@ -1907,51 +1907,75 @@ func (e *TrimExec) Next() (*Row, error) {
 // UnionExec has multiple source Executors, it executes them sequentially, and do conversion to the same type
 // as source Executors may has different field type, we need to do conversion.
 type UnionExec struct {
-	schema expression.Schema
-	Srcs   []Executor
-	ctx    context.Context
-	cursor int
+	schema   expression.Schema
+	Srcs     []Executor
+	ctx      context.Context
+	inited   bool
+	finished atomic.Value
+	rowsCh   chan *Row
+	wg       sync.WaitGroup
+	closedCh chan struct{}
+	errCh    chan error
 }
 
 // Schema implements the Executor Schema interface.
 func (e *UnionExec) Schema() expression.Schema {
 	return e.schema
 }
 
-// Next implements the Executor Next interface.
-func (e *UnionExec) Next() (*Row, error) {
+func (e *UnionExec) waitAllFinished() {
+	e.wg.Wait()
+	close(e.rowsCh)
+	close(e.closedCh)
+}
+
+func (e *UnionExec) fetchData(idx int) {
+	defer func() {
+		e.wg.Done()
+	}()
 	for {
-		if e.cursor >= len(e.Srcs) {
-			return nil, nil
+		if e.finished.Load().(bool) {
+			return
 		}
-		sel := e.Srcs[e.cursor]
-		row, err := sel.Next()
+		row, err := e.Srcs[idx].Next()
 		if err != nil {
-			return nil, errors.Trace(err)
+			e.finished.Store(true)
+			e.errCh <- err
+			return
 		}
 		if row == nil {
-			e.cursor++
-			continue
-		}
-		if e.cursor != 0 {
-			for i := range row.Data {
-				// The column value should be casted as the same type of the first select statement in corresponding position.
-				col := e.schema[i]
-				var val types.Datum
-				val, err = row.Data[i].ConvertTo(e.ctx.GetSessionVars().StmtCtx, col.RetType)
-				if err != nil {
-					return nil, errors.Trace(err)
-				}
-				row.Data[i] = val
-			}
+			return
 		}
+		e.rowsCh <- row
```

Comment:
If there are too many rows, the overhead of channel will be huge. It is better to use a batch way.